### DISCUSSION TITLE
Before in a forEach loop

### DISCUSSION THREAD

**Discussion by Ilie-c** [cypress-io/cypress#20934] [Questions and Help] [answered]
Hello,

I am studying cypress atm and I had a curiosity about a test. 
### The scenario: 
Based on the datasheet provided in the envoirment, in this case array of paths, I want to do in the same context, a suite of tests for every element of the array. Since I want a suite of tests I am in need of some before and/or beforeEatch to eliminate 'boiler' effect etc. 
Code example:
```
let array = Cypress.env("paths")
context('Verify navbar contents on all paths of app', function() {
    array.forEach((path)=>{
        before(()=>{
            cy.visit(path)
        })
        it('check navbar contains title for '+ path, () => {
            cy.get('.container').first().contains('cypress.io').should('exist')
        })
          it('show active class for current page '+ path, () => {
              cy.get('.container').first().find('[href="'+path+'"]').parent().should('have.class','active')
          })

    })
})
```

### The problem:
I use a before in a forEach function and since there is the async difference between js and cypress on every iteration the before will be runned for the entirety of the array:
![image](https://user-images.githubusercontent.com/89133734/161935005-77f3844e-63b4-40f7-a2b3-1a2b9cf3b1bd.png)
![image](https://user-images.githubusercontent.com/89133734/161935112-6be30734-caf2-4da6-81c1-bc2499aa65b9.png)
This means every time the active class will be the last one, that is the dropdown of 'commands' resulting in mostly test fail.

**I know that** if put the forEach outside of context it will work somewhat as intended. The issue will be that I'll have a big number of contexts based only on this test data....
![image](https://user-images.githubusercontent.com/89133734/161934365-5b3736aa-ade0-403f-aba2-ed282a0cd27a.png)

**Comment by jjhelguero** [comment, marked as answer]
It sounds like [`cypress-each`](https://github.com/bahmutov/cypress-each) would help you out here. You'll have the visit inside the `it` block and group all your tests together.

```js
import 'cypress-each'

const array = [
    ['/', 'Home page'],
    ['/about', 'About Page'],
    ['/contact', 'Contact Page']
]

describe('iterate and hit each page', () => {
    it.each(array)('%s page has %s header', (page, header) => {
        cy.visit(page)
        cy.contains('h1', header).should('be.visible')
    })
})
```

![it-each](https://user-images.githubusercontent.com/56001598/162018475-ee8f2239-45bd-4680-949b-bddfedb184a4.gif)

**Reply by Ilie-c** [reply]
Thank you, gonna give it a shot tomorrow and come back with an update.

**Reply by Ilie-c** [reply]
It did helped me, thanks!
Yet I wanted to understand things a bit more and found that:
In this case it will be equally good to use array.forEatch since is only one it:
```JS
context('Verify navbar contents on all paths of app using array.foreach', ()=> {
    array.forEach((path)=>{
        it('check navbar contains title for '+path, () => {
            cy.visit(path)
            cy.get('.container').first().contains('cypress.io').should('exist')
            if(path!='/'){
                cy.get('.container').first().find('[href="'+path+'"]').parent().should('have.class','active')
            }
        })
    })
})
```
### This library will help me in the future on other issues witch is great. Thank you

**Reply by muratkeremozcan** [reply]
100% this

Try a tutorial

https://dev.to/muratkeremozcan/crud-api-testing-a-deployed-service-with-cypress-using-cy-api-spok-cypress-data-session-cypress-each-4mlg
